### ISSUE TITLE
Listening function `.on` does not compile in Typescript when following example from README

### ISSUE THREAD

**Issue by vaughanj10** [supabase/realtime-js#157]
# Bug report

## Describe the bug

When referencing the example code in the README for my own usage of the library, the Typescript compiler fails to compile the format and usage of the `on` function. Example:

```
import { RealtimeClient } from "@supabase/realtime-js";

const client = new RealtimeClient("ws://realtime:6000/socket");

client.connect();
client.onOpen(() => console.log("[supabase service] Socket opened."));
client.onClose(() => console.log("[supabase service] Socket closed."));
client.onError((e) =>
  console.log("[supabase service] Socket error", e.message)
);

const locationTable = client.channel("realtime:public:location");
locationTable.on("INSERT", (e: any) => console.log(e));
locationTable.subscribe();
```

This line `locationTable.on("INSERT", (e: any) => console.log(e));` results in the Typescript error: 

```
error TS2345: Argument of type '(e: any) => void' is not assignable to parameter of type '{ [key: string]: string; } & Function'.
  Type '(e: any) => void' is not assignable to type '{ [key: string]: string; }'.
    Index signature for type 'string' is missing in type '(e: any) => void'.
```

When I switch to running this file as plain Javascript, everything works as expected.

## To Reproduce

Steps to reproduce the behavior, please provide code snippets or a repository:

1. Place some example code comparable to what I posted above in a Typescript index.ts file
2. Run with `ts-node index.ts`. 

## Expected behavior

I would expect the function signature of `.on` to remain usable and consistent whether using Javascript or Typescript.

## System information

- OS: Ubuntu
- Version of supabase-js: 1.6.2
- Version of Node.js: v16.14.0

**Comment by w3b6x9** [comment]
@vaughanj10 You imported `RealtimeClient` yet mentioned that your `supabase-js` version is 1.6.2. Either way, can you please upgrade to the latest version and try again.

**Comment by vaughanj10** [comment]
Tried again on 1.7.2. Same result. `'(e: any) => void' is not assignable to parameter of type '{ [key: string]: string; } & Function'.`

This type error seems pretty funky to me. Why would a type be a map of string -> strings AND a function? 

I think it has to do with the fact that when you create a channel, it's type is `RealtimeChannel | RealtimeSubscription`. We can see that `RealtimeSubscription` has a type for `.on` of:

```
  on(event: string, callback: Function)
```

and `RealtimeChannel` has a type of `.on` of:

```
  on(type: string, filter?: { [key: string]: string }, callback?: Function) {
```

Might it be that the typescript compiler is getting confused that the type of `.on` has different parameters types in a different in order? The second parameter for each function is a Function (for RealtimeSubscription) and a ` { [key: string]: string }` (for RealtimeChannel).

Here is an example that I just wrote:

```
class A {
  on(x: boolean, y: string) { return x }
}

class B {
  on(x: boolean, y: Function, z: string) { return x }
}

function getInstance(getA: boolean): A | B {
  if (getA) {
    return new A();
  } else {
    return new B();
  }
}

const instance = getInstance(true);
instance.on(true, () => {}, "")
```

If you run this code through ts-node, then you will receive the TypeScript error: `error TS2345: Argument of type '() => void' is not assignable to parameter of type 'Function & string'.` 
![image](https://user-images.githubusercontent.com/1255074/165137629-6ab9e5d9-d9d7-4b3f-a124-81e2718f405e.png)

This is pretty much the same situation that is encountered in the realtime-js library. I think the fix could just be to make sure that the callback function is the second parameter for both `.on` functions, or to find a better way infer whether the channel is actually just a `RealtimeChannel` or just a `RealtimeSubscription` and not either.

**Comment by molaeiali** [comment]
I have the same problem, there needs to be another override of the `.on` function without the filter.

When you have two optional parameters in a function, you cannot, not provide the first one

Another solution can be the filter type. it can be `{ [key: string]: string } | undefined` for example so that we can pass undefined instead of the filter.

**Comment by w3b6x9** [comment]
Sorry about this, everyone! Let me revert these changes and will ping here once it's released.

Planning some larger changes that will do a better job of this.

**Comment by w3b6x9** [comment]
This should be resolved after upgrading to `realtime-js` `v1.7.4`.
